### PR TITLE
feat: simplifies main navigation

### PR DESCRIPTION
**Pull request by kleinfreund**
Simplifies main navigation by removing the fly out navigation that used to contain the “Service Mesh” and “Diagnostics” items. The latter was included in the regular main navigation.

Turns on a setting for validating TypeScript inside Vue SFC templates and corrects a handful related issues.

Signed-off-by: Philipp Rudloff <philipp.rudloff@konghq.com>

### CODE REVIEW

**Comment by kleinfreund** [comment]
@lahabana Yes, we can do that. I’ve discarded this thought a while ago because we would likely (visually) hide the "Diagnostics" text and only show an icon, but then just the gear would be probably be taken for settings.

Looking at https://kongponents.konghq.com/components/icon.html, there isn't an immediately obvious icon available that signals "Diagnostics". Maybe "dashboard", but that looks more like a performance-related icon. "brain" could be an option.

This is what I currently have:

image

**Comment by kleinfreund** [comment]
(Messed something up in my branch management and accidentally force-pushed in what was evidently a rebased history)

**Comment by slonka** [comment]
The screenshots are old in the first post, new ones:

image

image

(I was searching for the diagnostics page, but then I read the comment it was moved)

**Review by slonka — APPROVED** [review]
It's really hard to review these big PRs. Haven't found any blockers though 👍

File: .eslintrc.js [hunk 1]
```diff
@@ -18,6 +18,7 @@
   extends: ['eslint:recommended', 'plugin:vue/vue3-recommended', 'standard', '@vue/typescript'],
   rules: {
     indent: 'off',
+    'func-call-spacing': 'off',
```
[the remark this answers is not in the document]
Comment (kleinfreund) [reply .eslintrc.js:21]: There is an extremely niche and specific reason why I disabled this rule:

```ts
const emit = defineEmits<{
  (event: 'mesh-selected', mesh: string): void
}>()
```

This code triggers this rule because, simply put, there _is_ a white space before the call operator.

I would like function calls to have no space before the call operator and function _definitions_ to have one because then I one can search all instances of _calls_ using `name(` while `name (` would only produce definitions.

I realized that I can simply turn on the related TypeScript-specific rule to avoid this error. So I’m doing the same as for the indent rule.

File: src/components/BreadcrumbsMenu.vue [hunk 2]
```diff
@@ -87,17 +87,14 @@ export default {
...
+      const { expandSidebar, ...cleanParams } = route.params
       const newParams = Object.assign({}, cleanParams, { mesh: null })
 
-      if (isMesh) {
-        return params.mesh
+      if (route.name === 'mesh-overview') {
+        return route.params.mesh
       } else {
-        return Object.values(newParams).filter((x) => x)[0]
+        const value = Object.values(newParams).filter((x) => x !== null)[0]
```
[the remark this answers is not in the document]
Comment (kleinfreund) [reply src/components/BreadcrumbsMenu.vue:96]: I find this part of the code pretty hard to parse when it comes to semantics in general. You’re right, I can simplify this with `find`. 👍

File: src/components/Sidebar/NavItem.vue [hunk 3]
```diff
@@ -2,278 +2,182 @@
...
 
-    &:hover {
-      background: none;
+      if (parentRoute.name === props.link) {
+        return true
+      }
+    } catch (error) {
+      if (error instanceof Error && error.message.includes('No match for')) {
+        // Unfortunately, `router.resolve` throws instead of returning `null` when a route can’t be resolved so we ignore this particular error because we don’t care about non-existing routes here.
+        console.warn(error)
```
[the remark this answers is not in the document]
Comment (kleinfreund) [reply src/components/Sidebar/NavItem.vue:130]: It provides a way for a developer to know that is a non-resolvable route. The navigation should be set-up using existing routes, but if somehow a non-existing route were to be introduced, this way, you’d at least see it in the console. I’m just making sure this logic doesn’t prevent the rendering of the navigation.

File: src/components/Utils/__snapshots__/EntityURLControl.spec.ts.snap [hunk 4]
```diff
@@ -50,7 +50,7 @@ exports[`EntityURLControl.vue renders snapshot 1`] = `
         </span>
         
         
-         Copy URL
+         Copy URL
```
[the remark this answers is not in the document]
Comment (kleinfreund) [reply src/components/Utils/__snapshots__/EntityURLControl.spec.ts.snap:53]: This is an expected consequence of how snapshots are generated in combination with how HTML renders white space. Consider the following code:

```html
<span>
  text node</span>
```

The HTML parser will render this as " text node" (side note: GitHub annoyingly `.trim()`s white space from code blocks so I can’t code-format this). This leads to snapshots reproducing this kind of whitespace because it accurately represents browsers would render.

File: src/store/index.ts [hunk 5]
```diff
@@ -93,7 +94,8 @@ export const storeConfig: StoreOptions<State> = {
     notifications,
     onboarding,
   },
-  state: initialState,
+  // Note: *Technically*, `initialState` here doesn’t hold all properties from `State`. All module state is missing. However, the module state MUST NOT be made optional in `State` as otherwise, it will be considered as optional when accessing, for example, `store.state.config.status`.
```
[the remark this answers is not in the document]
Comment (kleinfreund) [reply src/store/index.ts:97]: This going to be a longer explanation, so strap in. (Note: in the following, I’m calling `RootState` what in our app is simply called `State` to distinguish things)

For a store like ours, there is a `RootState` which is the type of all state that belongs to the store created with `createStore` in `@/store/store.ts`. This store _also_ has **namespaced** modules which introduce more state under named keys. Each module state also has a type (e.g. `ModuleState`).

With this, we have global state (e.g. `state.key`) and module state `state.moduleName.key`. The way Vuex sets up its types imply that `RootState` does not include module state. Taking this further: creating a store receives an initial store configuration (with the type of `RootState`. This operation returns a `store` object with a `state` property which _also_ has the type `RootState`. With what I explained above, this is not correct. the resulting store’s type _does_ include the module state (under the respective module names), but it doesn’t seem to be possible on Vuex’s side to correctly infer this type relation.

The way how I initialize our store is a compromise of this problem that ensures that we can access `store.state.moduleName.key` without the need to _always_ type-check whether `moduleName` exists. We know it does because we always add it.

I asked about a better way for this in the Vue discord where I provided the following minimal example illustrating the problem. It includes the two steps I also used in Kuma GUI to mitigate most of the problem (i.e. what you asked about).

---

Here is a minimal example illustrating the problem:

```typescript
import { createStore, Module, StoreOptions } from 'vuex'

const rootState = {
    key: 'value',
}

const testModuleState = {
    moduleKey: 'moduleValue',
}

/**
 * **Issue 1**: Defining the root state like this
 * makes it impossible to access module state with,
 * for example, `store.state.testModule.moduleKey`.
 * The resulting TypeScript error occurs:
 *
 * ~~~
 * store.state.testModule.moduleKey
 * //          ^^^^^^^^^^
 * // Property 'testModule' does not exist on type '{ key: string; }'.
 * ~~~
 *
 * That's because the type assigned to the result of calling `createStore`
 * does in fact not include state from modules.
 *
 * As a workaround, one can add the module state manually.
 */
type RootState = typeof rootState// & { testModule: TestModuleState }

type TestModuleState = typeof testModuleState

const testModule: Module<TestModuleState, RootState> = {
    namespaced: true,
    state: testModuleState,
}

const storeOptions: StoreOptions<RootState> = {
    /**
     * **Issue 2**: When adding, via intersection as shown above,
     * the module state to `RootState`,
     * the following TypeScript error occurs:
     *
     * ~~~
     *    state: rootState
     * // ^^^^^
     * // Type '{ key: string; }' is not assignable to type 'RootState | (() => RootState)'.
     * ~~~
     *
     * As a workaround, one can manually cast `state as RootState`,
     * but this loses you true TypeScript errors.
     */
    state: rootState/* as RootState */,
    modules: {
        testModule,
    },
}

const store = createStore<RootState>(storeOptions)

store.state.key
store.state.testModule.moduleKey
//          ^^^^^^^^^^
// Property 'testModule' does not exist on type '{ key: string; }'. ts(2339)
```

**Comment by kleinfreund** [comment]
> It's really hard to review these big PRs.

Sorry for this; it’s entirely my bad. I’ve accidentally pushed work (the big majority of the changes now in this PR) I’ve done on top of this PR’s branch while adding (trying to, anyway) the changes for moving the diagnostics item. 😭 Thank you for reviewing it anyway.

File: src/components/BreadcrumbsMenu.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (kleinfreund) [reply src/components/BreadcrumbsMenu.vue:96]: I’ve reworked this component’s logic to avoid 3 out of the 5 current cases and with that, removed the logic that makes use of this code.

File: src/store/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (kleinfreund) [reply src/store/index.ts:97]: And to actually directly answer the following question:

> why is it a problem that it's optional?

We would need to always type-guard access to module state which is **always** unnecessary because we know it exists as long as our state types are complete and accurate.

`store.state.config.status` always exists once we created a store, but we would need to at least white `store.state.config?.status` to read it and use an if statement to write to it (because you can’t use optional chaining on the left hand side of an assignment).

[the remark this answers is not in the document]
Comment (slonka) [reply src/store/index.ts:97]: jeez, this is a blogpost on it's own 😄 thanks for explaining!
